Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,95 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and optional type names that contain
+`public void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
+```
+
+The list is a semicolon-delimited list of assembly paths (absolute, or
+relative to the working directory), each with an optional type name
+(following an exclamation mark). If no type name is specified, the
+default is `StartupHook.Initializer`.
+
+Setting this environment variable will cause each of the specified
+types' `public void Initialize()` method to be called before the main
+assembly is loaded. The environment variable will be inherited by
+child processes by default. It is up to the `StartupHook.dll`s and
+user code to decide what to do about this - `StartupHook.dll` may
+clear them to prevent this behavior globally, if desired.
+
+Specifically, hostpolicy starts up coreclr and sets up a new AppDomain
+with each `StartupHook.dll` on the TPA list. It then invokes a private
+method in `System.Private.CoreLib`, which will call each
+`StartupHookType.Initialize()` in turn. This gives `StartupHookType` a
+chance to set up new `AssemblyLoadContext`s, or register other
+callbacks. After the `Initialize()` methods return, control returns to
+hostpolicy, which starts up the main entry point of the app like
```

Comment:
"which will then call the main entry..."

Reply:
Fixed!